File: Blind75/MaxSubArray.cpp
#include <vector>
using namespace std;

class Solution {
public:
  int maxSubArray(vector<int> &nums) {
    int n = nums.size();
    vector<int> dp(n, 0);
    dp[0] = nums[0];
    int maxSum = dp[0];

    for (int i = 1; i < n; i++) {
      dp[i] = nums[i] + max(0, dp[i - 1]);
      maxSum = max(maxSum, dp[i]);
    }
    return maxSum;
  }
};

File: Blind75/MaxSubArray.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        dp[0] = nums[0]
        maxValue = nums[0]

        for i in range(1, len(nums)):
            dp[i] = nums[i] + max(0, dp[i - 1])
            maxValue = max(maxValue, dp[i])

        return maxValue
